README
NAME
    Log::Dispatch::Message::Passing - log events to Message::Passing

SYNOPSIS
    In your application code:

      use Log::Dispatch;
      use Log::Dispatch::Message::Passing;
      use Message::Passing::DSL;

      my $log = Log::Dispatch->new;

      $log->add(Log::Dispatch::Message::Passing->new(
            name      => 'myapp_aggregate_log',
            min_level => 'debug',
            output    => log_chain {
                output zmq => (
                    class => 'ZeroMQ',
                    connect => 'tcp://192.168.0.1:5558',
                );
            },
      ));

      $log->warn($_) for qw/ foo bar baz /;

    On your central log server:

      message-pass --input ZeroMQ --input_options '{"socket_bind":"tcp://*:5558"}' --output File --output_options '{"filename":"myapp_aggregate.log"}'

DESCRIPTION
    This provides a Log::Dispatch log output system that sends logged events
    to Message::Passing.

    This allows you to use any of the Message::Passing emitters or filters
    to process log events and send them across the network, and you can use
    the toolkit to trivially construct a log aggregator.

METHODS
  "new"
     my $table_log = Log::Dispatch::Message::Passing->new(\%arg);

    This method constructs a new Log::Dispatch::Message::Passing output
    object.

    Required arguments are:

      output - a L<Message::Passing> L<Output|Message::Passing::Role::Output> class.

  log_message
    This is the method which performs the actual logging, as detailed by
    Log::Dispatch::Output.

SEE ALSO
    Message::Passing
        The logging framework itself, allowing you to very simply build log
        aggregation and processing servers.

    Message::Passing::Output::ZeroMQ
        The recommended network protocol for aggregating or transporting
        messages across the network.

        Note that whilst this transport is recommended, it is NOT required
        by this module, so you need to require (and depend on)
        Message::Passing::ZeroMQ separately.

    example/ directory
        Instantly runnable SYNOPSIS - plug into your application for easy
        log aggregation.

AUTHOR
    Tomas Doran (t0m) "<bobtfish@bobtfish.net>"

SPONSORSHIP
    This module exists due to the wonderful people at Suretec Systems Ltd.
    <http://www.suretecsystems.com/> who sponsored its development for its
    VoIP division called SureVoIP <http://www.surevoip.co.uk/> for use with
    the SureVoIP API -
    <http://www.surevoip.co.uk/support/wiki/api_documentation>

COPYRIGHT
    Copyright Suretec Systems Ltd. 2012.

LICENSE
    GNU Affero General Public License, Version 3

    If you feel this is too restrictive to be able to use this software,
    please talk to us as we'd be willing to consider re-licensing under less
    restrictive terms.



Changes
0.008
   - Work with upcoming versions of Log::Dispatch

0.007
   - Fix test - newer versions of Log::Dispatch
     return 'warning' instead of 'warn'

0.006
   - Fix broken dep in Makefile.PL RT#78799

0.005
    - More doc and example fixes.

0.004
    - Doc and example fixes

0.003
    - Rename

0.002
    - Better docs
    - Better testing

0.001
    - Initial version



Makefile.PL
use strict;
use warnings;
use inc::Module::Install;
use Module::Install::Authority;
use Module::Install::AuthorRequires;
use Module::Install::AuthorTests;

all_from 'lib/Log/Dispatch/Message/Passing.pm';

license 'AGPL_3';

authority('GHENRY');
resources(
    repository => "git://github.com/suretec/Log-Dispatch-Message-Passing.git",
);

requires 'Message::Passing' => '0.006';
requires 'Log::Dispatch';
requires 'Scalar::Util';

test_requires 'Test::More' => '0.88';

author_requires 'Log::Message::Structured';
author_requires 'Test::Pod' => '1.14';
author_requires 'Test::NoTabs';
author_requires 'Test::Pod::Coverage' => '1.04';
author_requires 'Pod::Coverage' => '0.19';
author_requires 'Pod::Coverage::TrustPod';
author_requires 'Test::Spelling';

author_tests 't/author';

if ($Module::Install::AUTHOR) {
    system("pod2text lib/Log/Dispatch/Message/Passing.pm > README")
        and die $!;
}

WriteAll;



lib/Log/Dispatch/Log/Stash.pm
package Log::Dispatch::Log::Stash;
use base qw(Log::Dispatch::Output);

use warnings;
use strict;
use Scalar::Util qw/ blessed /;
use Carp qw/ confess /;

our $VERSION = '0.001';

sub new {
  my ($class, %arg) = @_;
  confess("Need an 'output' argument") unless $arg{output};
  my $output = $arg{output};
  confess("output => $output must be an object which can ->consume")
    unless blessed($output) && $output->can('consume');

  my $self = { output => $output };

  bless $self => $class;

  # this is our duty as a well-behaved Log::Dispatch plugin
  $self->_basic_init(%arg);

  return $self;
}

sub log_message {
  my ($self, %p) = @_;
  $self->{output}->consume({%p});
}

=head1 NAME

Log::Dispatch::Log::Stash - log events to Log::Stash

=head1 SYNOPSIS

In your application code:

  use Log::Dispatch;
  use Log::Dispatch::Log::Stash;
  use Log::Stash::DSL;

  my $log = Log::Dispatch->new;

  $log->add(Log::Dispatch::Log::Stash->new(
        name      => 'myapp_aggregate_log',
        min_level => 'debug',
        output    => log_chain {
            output zmq => (
                class => 'ZeroMQ',
                connect => 'tcp://192.168.0.1:5558',
            );
        },
  ));

  $log->warn($_) for qw/ foo bar baz /;

On your central log server:

  logstash --input ZeroMQ --input_options '{"socket_bind":"tcp://*:5558"}' --output STDOUT > myapp_aggregate.log

=head1 DESCRIPTION

This provides a Log::Dispatch log output system that sends logged events to
L<Log::Stash>.

This allows you to use any of the Log::Stash emitters or filters
to process log events and send them across the network, and you can use
the toolkit to trivially construct a log aggregator.

=head1 METHODS

=head2 C<< new >>

 my $table_log = Log::Dispatch::Log::Stash->new(\%arg);

This method constructs a new Log::Dispatch::Log::Stash output object.

Required arguments are:

  output - a L<Log::Stash> OUTPUT class.

=head2 log_message

This is the method which performs the actual logging, as detailed by
Log::Dispatch::Output.

=cut

=head1 SEE ALSO

=over

=item L<Log::Stash>

The logging framework itself, allowing you to very simply build log
aggregation and processing servers.

=item L<Log::Stash::Output::ZeroMQ>

The recommended network protocol for aggregating or transporting messages
across the network.

Note that whilst this transport is recommended, it is B<NOT> required by
this module, so you need to require (and depend on) L<Log::Stash::ZeroMQ>
separately.

=item example/ directory

Instantly runnable SYNOPSIS - plug into your application for easy log
aggregation.

=back

=head1 AUTHOR

Tomas Doran (t0m) C<< <bobtfish@bobtfish.net> >>

=head1 SPONSORSHIP

This module exists due to the wonderful people at Suretec Systems Ltd.
<http://www.suretecsystems.com/> who sponsored it's development for its
VoIP division called SureVoIP <http://www.surevoip.co.uk/> for use with
the SureVoIP API - 
<http://www.surevoip.co.uk/support/wiki/api_documentation>

=head1 COPYRIGHT

Copyright Suretec Systems Ltd. 2012.

=head1 LICENSE

GNU Affero General Public License, Version 3

If you feel this is too restrictive to be able to use this software,
please talk to us as we'd be willing to consider re-licensing under
less restrictive terms.

=cut

1;



lib/Log/Dispatch/Message/Passing.pm
package Log::Dispatch::Message::Passing;
use base qw(Log::Dispatch::Output);

use warnings;
use strict;
use Scalar::Util qw/ blessed /;
use Carp qw/ confess /;

our $VERSION = '0.008';

sub new {
  my ($class, %arg) = @_;
  confess("Need an 'output' argument") unless $arg{output};
  my $output = $arg{output};
  confess("output => $output must be an object which can ->consume")
    unless blessed($output) && $output->can('consume');

  my $self = { output => $output };

  bless $self => $class;

  # this is our duty as a well-behaved Log::Dispatch plugin
  $self->_basic_init(%arg);

  return $self;
}

sub log_message {
  my ($self, %p) = @_;
  if ($p{level} =~ /^\d+$/) {
      require Log::Dispatch::Util;
      $p{level} = $Log::Dispatch::Util::LevelNames[ $p{level} ];
  }
  $self->{output}->consume({%p});
}

=head1 NAME

Log::Dispatch::Message::Passing - log events to Message::Passing

=head1 SYNOPSIS

In your application code:

  use Log::Dispatch;
  use Log::Dispatch::Message::Passing;
  use Message::Passing::Filter::Encoder::JSON;
  use Message::Passing::Output::ZeroMQ;

  my $log = Log::Dispatch->new;

  $log->add(Log::Dispatch::Message::Passing->new(
        name      => 'myapp_aggregate_log',
        min_level => 'debug',
        output    => Message::Passing::Filter::Encoder::JSON->new(
            output_to => Message::Passing::Output::ZeroMQ->new(
                connect => 'tcp://192.168.0.1:5558',
            ),
        ),
  ));

  $log->warn($_) for qw/ foo bar baz /;

On your central log server:

  message-pass --input ZeroMQ --input_options '{"socket_bind":"tcp://*:5558"}' \
    --output File --output_options '{"filename":"myapp_aggregate.log"}'

=head1 DESCRIPTION

This provides a L<Log::Dispatch> log output system that sends logged events to
L<Message::Passing>.

This allows you to use any of the Message::Passing outputs or filters
to process log events and send them across the network, and you can use
the toolkit to trivially construct a log aggregator.

=head1 METHODS

=head2 C<< new >>

 my $table_log = Log::Dispatch::Message::Passing->new(\%arg);

This method constructs a new Log::Dispatch::Message::Passing output object.

Required arguments are:

  output - a L<Message::Passing> L<Output|Message::Passing::Role::Output> class.

=head2 log_message

This is the method which performs the actual logging, as detailed by
Log::Dispatch::Output.

=cut

=head1 SEE ALSO

=over

=item L<Message::Passing>

The logging framework itself, allowing you to very simply build log
aggregation and processing servers.

=item L<Message::Passing::Output::ZeroMQ>

The recommended network protocol for aggregating or transporting messages
across the network.

Note that whilst this transport is recommended, it is B<NOT> required by
this module, so you need to require (and depend on) L<Message::Passing::ZeroMQ>
separately.

=item example/ directory

Instantly runnable SYNOPSIS - plug into your application for easy log
aggregation.

=back

=head1 AUTHOR

Tomas Doran (t0m) C<< <bobtfish@bobtfish.net> >>

=head1 SPONSORSHIP

This module exists due to the wonderful people at Suretec Systems Ltd.
<http://www.suretecsystems.com/> who sponsored it's development for its
VoIP division called SureVoIP <http://www.surevoip.co.uk/> for use with
the SureVoIP API - 
<http://www.surevoip.co.uk/support/wiki/api_documentation>

=head1 COPYRIGHT

Copyright Suretec Systems Ltd. 2012.

=head1 LICENSE

GNU Affero General Public License, Version 3

If you feel this is too restrictive to be able to use this software,
please talk to us as we'd be willing to consider re-licensing under
less restrictive terms.

=cut

1;

